### PR TITLE
test(cache): fix sync test error type, add concurrency tests, remove duplicate

### PR DESCRIPTION
**Pull request by trevor-vaughan**
TestSync_FailureOnMissingPolicy previously passed for the wrong reason:
the mock returned a plain error instead of registry.ErrVersionNotFound,
causing it to hit the "registry unreachable" branch instead of the
"not found" branch. Fix the mock to return the correct error type.

Add TestSync_RegistryUnreachableError to cover the "registry unreachable"
branch with a non-ErrVersionNotFound error via the new SetError method.

Replace the sequential TestSync_StressConcurrentFailures with two real
concurrent tests (ConcurrentDifferentPolicies and ConcurrentMixedFailures)
that use goroutines and sync.WaitGroup to catch data races under -race.

Fix TestSync_SHA384Digest to seed the mock with the sha384 lookup ref
so sync succeeds, and assert NoError instead of silently discarding it.

Fix isDuplicateErr to use strings.Contains instead of exact match,
since OCI store errors include digest and media type context.

Remove TestComplypackSync_UnsignedWarning which was functionally identical
to TestComplypackSync_FetchAndStore and tested nothing about unsigned
warnings.

Refs: #653

Assisted-by: Claude Opus 4.6
Signed-off-by: Trevor Vaughan <tvaughan@redhat.com>

### CODE REVIEW

File: internal/cache/sync_test.go [hunk 1]
```diff
@@ -187,67 +218,133 @@ func TestSync_RedownloadAfterDeletion(t *testing.T) {
...
+	var wg gosync.WaitGroup
+	wg.Add(workers)
 
-	for i := 0; i < iterations; i++ {
-		state, loadErr := cache.LoadState(cacheDir)
-		require.NoError(t, loadErr, "state load must not fail on iteration %d", i)
+	for i := 0; i < workers; i++ {
+		go func(workerID int) {
+			defer wg.Done()
+			state, loadErr := cache.LoadState(cacheDir)
```
Comment (marcusburghardt) [inline internal/cache/sync_test.go]: **[HIGH] Race condition on `state.json` causes CI failure**

All 10 goroutines share the same `cacheDir` and each calls `cache.LoadState(cacheDir)` inside the goroutine. When `SyncPolicy` completes, it calls `SaveState()` → `os.WriteFile(statePath, data, 0600)`, which truncates the file before writing. A concurrent `LoadState` from another goroutine observes the truncated file and fails with `unexpected end of JSON input`.

CI log: `worker 6: state load failed: failed to parse state file .../state.json: unexpected end of JSON input`

`TestSync_ConcurrentMixedFailures` has the same structural pattern but is less likely to trigger because odd-numbered workers fail at `DefinitionVersion` (before `SaveState`), reducing concurrent writers from 10 to ~5.

Suggested fixes (any one resolves the failure):

1. **Per-worker cache directories** (simplest): give each goroutine its own `cacheDir` subdirectory. Tests concurrent mock access without file-level contention.
2. **Pre-load shared state, serialize saves**: load state once before spawning, share `*State`, add a mutex around `UpdatePolicyState` + `SaveState`.
3. **Atomic writes in `SaveState`** (production code change): write-then-rename pattern. Fixes the root cause but changes production scope — likely a separate PR per Constitution III.

**Review by marcusburghardt — CHANGES_REQUESTED** [review]
Test quality improvements for `internal/cache/` — fixes mock error type, adds concurrency tests, removes duplicate test. The `isDuplicateErr` fix, `TestSync_RegistryUnreachableError` addition, SHA384 test fix, and duplicate removal are all correct and improve test fidelity.

However, `TestSync_ConcurrentDifferentPolicies` has a race condition on `state.json` file I/O that causes the `unit-test` CI check to fail. See inline comment for details and suggested fixes.

_This review was generated by /review-pr (AI-assisted)._

**Review by marcusburghardt — DISMISSED** [review]
LGTM

File: internal/cache/cachetest/mock_source.go [hunk 2]
```diff
@@ -198,5 +221,5 @@ func isDuplicateErr(err error) bool {
 	if err == nil {
 		return false
 	}
-	return err.Error() == "already exists"
+	return strings.Contains(err.Error(), "already exists")
```
Comment (sonupreetam) [inline internal/cache/cachetest/mock_source.go]: **[HIGH] Use `errors.Is` instead of `strings.Contains` for error matching (CS-006)**

The old exact match never worked because `ocistore.Store.Push()` wraps `errdef.ErrAlreadyExists` with digest and media type context (e.g., `"sha256:abc...: application/vnd.oci.empty.v1+json: already exists"`). `strings.Contains` fixes the immediate problem but is fragile -- it relies on the error message text remaining stable across ORAS versions.

Every other caller in the dependency tree (6 sites across oras-go, go-gemara, complypack) uses `errors.Is`:

```suggestion
	return errors.Is(err, errdef.ErrAlreadyExists)
```

Requires adding `"errors"` to the import block (it's already imported as `strings` is) and `errdef "oras.land/oras-go/v2/errdef"` to the import section.

File: internal/cache/cachetest/mock_source.go [hunk 3]
```diff
@@ -54,16 +69,24 @@ func (m *MockPolicySource) SeedPolicy(policyID, version, digestStr string) {
...
 func (m *MockPolicySource) DefinitionVersion(_ context.Context, lookupRef string) (string, string, error) {
 	m.mu.Lock()
 	m.LastLookupRef = lookupRef
 	m.mu.Unlock()
 
 	m.mu.RLock()
 	defer m.mu.RUnlock()
+
+	if injected, ok := m.errors[lookupRef]; ok {
+		return "", "", injected
```
Comment (sonupreetam) [inline internal/cache/cachetest/mock_source.go:85]: **[MEDIUM] Mock error format diverges from production, causing double-wrapped "policy" prefix**

The production `DefinitionVersion` (via `fetchVersion` in `registry/client.go:138`) returns `fmt.Errorf("%w: %s tag %q", ErrVersionNotFound, ref, tag)` -- the sentinel first, then context. `SyncPolicy` then wraps with `fmt.Errorf("policy %s: %w", policyID, err)`.

This mock wraps as `fmt.Errorf("policy %s: %w", lookupRef, ErrVersionNotFound)`, so `SyncPolicy` produces `"policy X: policy X: version not found in registry"` -- redundant prefix.

Consider matching the production format or simplifying to just `fmt.Errorf("%w", registry.ErrVersionNotFound)`, which keeps mock errors clean and avoids simulating formatting the mock doesn't need to test.

**Review by sonupreetam — CHANGES_REQUESTED** [review]
Good fix -- `TestSync_FailureOnMissingPolicy` was indeed exercising the wrong error branch. The `ErrVersionNotFound` wrapping, `SetError` addition, concurrency test rewrite, and SHA384 seeding fix are all solid improvements.

Two items on code introduced in this PR:

1. `isDuplicateErr` should use `errors.Is(err, errdef.ErrAlreadyExists)` instead of `strings.Contains` -- matches every other caller in the dependency tree and is immune to message format changes.

2. The mock's error format `fmt.Errorf("policy %s: %%w", ...)` diverges from production, causing `SyncPolicy` to produce a double-wrapped `"policy X: policy X: ..."` message.

_This review was generated by /review-pr (AI-assisted)._

**Review by sonupreetam — DISMISSED** [review]
LGTM

**Review by marcusburghardt — DISMISSED** [review]
LGTM

**Comment by marcusburghardt** [comment]
@trevor-vaughan there were two PRs editing the `.gitattributes` file slightly different. The merged one is cleaner so we can remove the `.gitattributes` file here to resolve the conflict.

**Comment by trevor-vaughan** [comment]
> @trevor-vaughan there were two PRs editing the `.gitattributes` file slightly different. The merged one is cleaner so we can remove the `.gitattributes` file here to resolve the conflict.

The merged one will result in not recursing and having issues. This one fixes that.

**Comment by marcusburghardt** [comment]
> > @trevor-vaughan there were two PRs editing the `.gitattributes` file slightly different. The merged one is cleaner so we can remove the `.gitattributes` file here to resolve the conflict.
> 
> The merged one will result in not recursing and having issues. This one fixes that.

Unfortunately I this is not what the official documentation says and can be confirmed with simple tests.
Both patterns are functionally identical in `.gitattributes`. Git's pattern matching follows `gitignore(5)` rules (per [`gitattributes(5)`](https://git-scm.com/docs/gitattributes)):

> The rules by which the pattern matches paths are the same as in `.gitignore` files

And [`gitignore(5)`](https://git-scm.com/docs/gitignore#_pattern_format) defines:

> A trailing `"/**"` matches everything inside. For example, `"abc/**"` matches all files inside directory `"abc"`, relative to the location of the `.gitignore` file, **with infinite depth**.

### Reproduction

```bash
cd /tmp && rm -rf gitattr-proof && mkdir gitattr-proof && cd gitattr-proof && git init -q
mkdir -p vendor/a/b/c
echo x > vendor/top.go && echo x > vendor/a/mid.go && echo x > vendor/a/b/c/deep.go
printf '* text eol=lf\nvendor/** -text\n' > .gitattributes && git add .
echo "--- vendor/** ---" && git check-attr text -- vendor/top.go vendor/a/mid.go vendor/a/b/c/deep.go
sed -i 's|vendor/\*\*|vendor/**/*|' .gitattributes && git add .gitattributes
echo "--- vendor/**/* ---" && git check-attr text -- vendor/top.go vendor/a/mid.go vendor/a/b/c/deep.go
rm -rf /tmp/gitattr-proof
```

Output is identical for both patterns at all nesting levels. Now that this is clarified, lets update the PRs and move forward. Thanks!

**Review by marcusburghardt — CHANGES_REQUESTED** [review]
https://github.com/complytime/complyctl/pull/663#issuecomment-4923165138

**Comment by trevor-vaughan** [comment]
@marcusburghardt You're absolutely right about the path matching. I was misremembering :face_exhaling:.

It turns out that we both missed something that I hadn't been aware of. Apparently, the best form to use is `vendor/** linguist-generated -diff -text` which will correct it both in GitHub scans and locally. Fixed in the rebased update.

**Review by sonupreetam — APPROVED** [review]
LGTM

**Comment by marcusburghardt** [comment]
> @marcusburghardt You're absolutely right about the path matching. I was misremembering 😮‍💨.
> 
> It turns out that we both missed something that I hadn't been aware of. Apparently, the best form to use is `vendor/** linguist-generated -diff -text` which will correct it both in GitHub scans and locally. Fixed in the rebased update.

I see @trevor-vaughan . I was even checking this morning the documentation about `linguist-generated` and `linguist-vendored` which were also both unfamiliar to me. It seems the `linguist-vendored` would fit better in this case but it would not hide vendored files in diff, which seems the main delta with `linguist-generated`. Considering we have other guardrails such as OSV-Scanner, Trivy, CodeQL we should be fine to hide the diff.

As a retrospective, the main focus of this PR were the tests and not the `.gitattributes`. Things were improved but it would be great if we could agree to separate these changes and keep the PRs focused so we can avoid similar situations and merge the PRs more quickly with less friction. Are we good? : )

**Review by marcusburghardt — APPROVED** [review]
LGTM

**Comment by marcusburghardt** [comment]
Just rebased using the web interface. This will preserve the previous approvals.
